pandapipes/component_models/flow_control_component.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2020-2021 by Fraunhofer Institute for Energy Economics
+# and Energy System Technology (IEE), Kassel. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+import numpy as np
+from numpy import dtype
+
+from pandapipes.component_models.junction_component import Junction
+from pandapipes.component_models.abstract_models import BranchWZeroLengthComponent, get_fluid
+from pandapipes.idx_branch import D, AREA, TL, JAC_DERIV_DP, JAC_DERIV_DP1, JAC_DERIV_DV, VINIT, \
+    RHO, LOAD_VEC_BRANCHES, ELEMENT_IDX
+from pandapipes.pf.result_extraction import extract_branch_results_without_internals
+
+
+class FlowControlComponent(BranchWZeroLengthComponent):
+    """
+
+    """
+
+    @classmethod
+    def table_name(cls):
+        return "flow_control"
+
+    @classmethod
+    def active_identifier(cls):
+        return "in_service"
+
+    @classmethod
+    def from_to_node_cols(cls):
+        return "from_junction", "to_junction"
+
+    @classmethod
+    def get_connected_node_type(cls):
+        return Junction
+
+    @classmethod
+    def create_pit_branch_entries(cls, net, branch_pit):
+        """
+        Function which creates pit branch entries with a specific table.
+        :param net: The pandapipes network
+        :type net: pandapipesNet
+        :param branch_pit:
+        :type branch_pit:
+        :return: No Output.
+        """
+        fc_pit = super().create_pit_branch_entries(net, branch_pit)
+        fc_pit[:, D] = net[cls.table_name()].diameter_m.values
+        fc_pit[:, AREA] = fc_pit[:, D] ** 2 * np.pi / 4
+        fc_pit[:, VINIT] = net[cls.table_name()].controlled_mdot_kg_per_s.values / \
+            (fc_pit[:, AREA] * fc_pit[:, RHO])
+
+    @classmethod
+    def adaption_before_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_lookups, options):
+        pass
+
+    @classmethod
+    def adaption_after_derivatives_hydraulic(cls, net, branch_pit, node_pit, idx_lookups, options):
+        # set all pressure derivatives to 0 and velocity to 1; load vector must be 0, as no change
+        # of velocity is allowed during the pipeflow iteration
+        f, t = idx_lookups[cls.table_name()]
+        fc_pit = branch_pit[f:t, :]
+        in_service_elm = np.isin(net[cls.table_name()].index.values,
+                                 fc_pit[:, ELEMENT_IDX].astype(np.int32))
+        active = net[cls.table_name()].control_active.values[in_service_elm]
+        fc_pit[active, JAC_DERIV_DP] = 0
+        fc_pit[active, JAC_DERIV_DP1] = 0
+        fc_pit[active, JAC_DERIV_DV] = 1
+        fc_pit[active, LOAD_VEC_BRANCHES] = 0
```

Comment:
This I do not get. I agree that the velocity is not changing. But LOAD_VEC_BRANCHES seems to me not a differential. Thus I would say, LOAD_VEC_BRANCHES is controlled_mdot_kg_per_s?

Reply:
I always stumble over this part as well... the LOAD_VEC_BRANCH is the difference between the calculated pressure drop and the pressure drop given by the current estimates of the pressure at the connected nodes. Whereas the LOAD_VEC_NODES is the part that resumes over the mass flows coming into each node. So we assume that the difference is 0, because we accept any assumed pressure values at the connected nodes to be calculated correctly, we assume that the flow controller can produce any pressure drop / lift, that is required to create the desired mass flow.

Reply:
Ahh, true, now I get it. Thanks!